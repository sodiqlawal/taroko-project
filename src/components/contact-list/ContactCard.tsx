"use client"
import { TOAST_OPTION } from "@/constant";
import useModal from "@/hooks/useModal";
import { useToast } from "@/hooks/useToast";
import { deleteContact } from "@/lib/api/contact";
import { Contact } from "@/models/contact";
import { isAxiosError } from "axios";
import Link from "next/link";
import { FC, useState } from "react";
import { FaUser } from "react-icons/fa"
import Button from "../miscellaneous/form/Button";
import Modal from "../modals";
import styles from "./contact-list.module.scss";

const ContactCard: FC<{ contact: Contact; refetch:() => void }> = ({ contact, refetch }) => {
    const deleteModal = useModal();
    const {toast} = useToast();
    const [loading, setLoading] = useState(false);

    const deleteAction = async () => {
        setLoading(true);
        try {
            await deleteContact(contact.id || "");

            toast({
                type: TOAST_OPTION.SUCCESS,
                message: "Contact deleted successfully"
            })

            // refetch all contacts 
            refetch();

            setLoading(false);

            deleteModal.close();

        } catch (error) {
            let message = isAxiosError(error) ? error.response?.data?.message || "" : error
            toast({
                type: TOAST_OPTION.ERROR,
                message
            })
            setLoading(false)
        }
    }

    if (!contact) return null;
    return (
        <div className={styles.card}>

            <div className={styles.top}>
                <div>
                    <div className={styles.user}>
                        <FaUser size={30} />
                        <p>{`${contact.first_name} ${contact.last_name}`}</p>
                    </div>

                    <p className={styles.job}><b>Job:</b> {contact.job}</p>

                </div>

                <div className={styles.cta}>
                    <Link href={`/${contact.id}/edit-contact`}>
                        <Button content="Edit" className={styles.button} simple />
                    </Link>
                    <Button content="Delete" onClick={deleteModal.open} />
                </div>

            </div>


            <p className={styles.desc}><b>Description:</b> {contact.description}</p>

            {/* this modal is triggered when we click on delete card button */}
            <Modal controller={deleteModal} title="Delete Card">
                <p className={styles.delete_desc}>Are you sure you want to delete this card ?</p>
                <div className={styles.button_cover}>
                    <Button content="No" isMedium isDanger onClick={deleteModal.close} growth />
                    <Button content="Yes" isMedium isPrimary growth onClick={deleteAction} loading={loading} disabled={loading} />
                </div>
            </Modal>

        </div>
    )
}

export default ContactCard;