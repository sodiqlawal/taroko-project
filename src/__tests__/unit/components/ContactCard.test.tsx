import { fireEvent, render, screen } from "@testing-library/react";
import ContactCard from "@/components/contact-list/ContactCard";


const defaultProps = {
    contact: {
        first_name: "Sodiq",
        last_name: "Lawal",
        job: "Software Engineer",
        description: "Tech savvy with business acument"
    },
    refetch: jest.fn
};


const setupContactCard = (props = defaultProps) =>
    render(
        <ContactCard {...props} />
    );

describe("The <ContactCard /> component", () => {
    it("should match snapshot", () => {
        const { asFragment } = setupContactCard();

        expect(asFragment()).toMatchSnapshot()
    });

    it("should display contact data", () => {
        setupContactCard();

        const {first_name, last_name, job, description} = defaultProps.contact;

        expect(screen.getByText(`${first_name} ${last_name}`)).toBeInTheDocument();
        expect(screen.getByText(job)).toBeInTheDocument();
        expect(screen.getByText(description)).toBeInTheDocument();
    });

    it("should not show delete alert modal on mount", () => {
        setupContactCard();

        expect(screen.queryByTestId(/modal/)).not.toBeInTheDocument();
    });

    it("should show delete alert modal on clicking delete button", () => {
        setupContactCard();

        const deleteBtn = screen.getByRole("button", { name: /Delete/i });

        fireEvent.click(deleteBtn)

        expect(screen.getByTestId(/modal/)).toBeInTheDocument();
    });

    it("should hide delete alert modal on clicking `No` on the modal", () => {
        setupContactCard();

        const deleteBtn = screen.getByRole("button", { name: /Delete/i });
        
        fireEvent.click(deleteBtn);
        
        const noBtn = screen.getByRole("button", { name: /No/i });
        
        fireEvent.click(noBtn)


        expect(screen.queryByTestId(/modal/)).not.toBeInTheDocument();
    });

});
