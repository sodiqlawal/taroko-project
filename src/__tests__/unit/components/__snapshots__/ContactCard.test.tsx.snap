// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <ContactCard /> component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="top"
    >
      <div>
        <div
          class="user"
        >
          <svg
            fill="currentColor"
            height="30"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
            />
          </svg>
          <p>
            Sodiq Lawal
          </p>
        </div>
        <p
          class="job"
        >
          <b>
            Job:
          </b>
           Software Engineer
        </p>
      </div>
      <div
        class="cta"
      >
        <a
          href="/undefined/edit-contact"
        >
          <button
            class="button button simple"
            type="button"
          >
            Edit
          </button>
        </a>
        <button
          class="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <p
      class="desc"
    >
      <b>
        Description:
      </b>
       Tech savvy with business acument
    </p>
  </div>
</DocumentFragment>
`;
